### ISSUE TITLE
Accessibility - CardActions not supporting alt text for images and buttons.

### ISSUE THREAD

**Issue by murcinar** [microsoft/BotFramework-WebChat#2675]
## Screenshots

![image](https://user-images.githubusercontent.com/58231579/70285350-0f66cb80-177c-11ea-9a68-164877efc52e.png)

## Version

Microsoft Bot Framework v4:
Microsoft.Bot.Builder 4.6.2 Nuget package

## Describe the bug

Through Accessibility Scan, we have received the following issues where we could not found a solution to implement using Bot Framework PromptOptions:
**Issue description: Ensures  elements have alternate text or a role of none or presentation
Issue description: Ensures buttons have discernible text**
These issues occurred where we have created suggested action items (like thumbs up/down, create a ticket, etc) using the **PromptOptions** class in Microsoft.Bot.Builder.Dialogs. Using the PromptActions, we see that there is no way to add alt text to an image or a button.
Can you please advise on how to achieve accessibility on the PromptOptions in Web Chat?

[Bug]

**Comment by corinagum** [comment]
https://github.com/microsoft/BotFramework-WebChat/pull/2613

Using PromptOptions doesn't have a direct impact on the accessibility of what's rendered in Web Chat. A PR has already been merged to make Suggested Actions accessible, and will be a part of the December 10 release.

**Comment by murcinar** [comment]
Today I have upgraded the Bot Framework to version 4.7.1 and when I deployed the changes and ran the Accessibility Insights tests again, I saw the same errors coming up again:
 
TestContext Messages:
Issue description: Ensures buttons have discernible text found in the following code:
 Issue description: Ensures <img> elements have alternate text or a role of none or presentation found in the following code:
 <button type="button"><img src="/images/thumbs-up-icon@2x.png"><nobr> </nobr></button>
 <button type="button"><img src="/images/thumbs-down-icon@2x.png"><nobr> </nobr></button>
 <img src="/images/thumbs-up-icon@2x.png">
 <img src="/images/thumbs-down-icon@2x.png">
 <img src="/images/chat-with-agent-icon@2x.png">
 <img src="/images/create-ticket-icon@2x.png">

Can you please comment on this if the fix was supposed to add alt text to the images or provide a property to accept alt text for the images and buttons? Like a new property on PromptOptions class?

**Comment by corinagum** [comment]
Is <TestContext> a component you have created? Can you share that code?

**Comment by murcinar** [comment]
Sure, it is pretty much as following:
private PromptOptions promptOption;
promptOption.Prompt.SuggestedActions.Actions.Add(ThumbsUpAction.Create(Images.ThumbsUpIconPath));
//similarly for thumbs down and etc..

ThumbsUpAction.Create returns the following:
return new CardAction(type: ActionTypes.PostBack,
                                  title: " ",
                                  value: CommonActionPayload.Create(ThumbsUpIntent),
                                  image: buttonImageUrl);

And I still don't see any property where we can pass any sort of alt text into the image/card action/ Prompt Options.

**Comment by corinagum** [comment]
Hmm, I see what you mean about the image not having alt text. Normally, the title (button text) functions as the label, but your button doesn't have a title.

However, I don't see the accessible solution implemented by Web Chat working for your case. The role should be set to none for the image, since it is supplemental to the title of the button. However, since your promptOption has no title, the button will remain inaccessible in your bot's case. 

Provided we implement the fix where we add an empty `alt` or `role` of none, I think it would still be necessary for you to use attachment middleware on suggestedActions to add an alt to the action if there is no title text. Regardless of when Web Chat does implement the fix, the modification for your bot in particular would remain the same. :) 

Thanks for bringing this back to my attention.

**Comment by murcinar** [comment]
Thanks for your help! By when do you think we can expect to have a fix for this?

**Comment by murcinar** [comment]
Also, I have found out another issue where on the CardAction class, there is no way to provide either an alt text or an aria-label. Title is the only field that goes in the button as a text but we already have an image that specifies the button so we don't want any text on the button. But from the accessibility stand point, screen readers like narrator announce "button" only in this case and does not specify what the button does. We need a way to be able to pass an aria-label to be accessibility compliant. Can you please prioritize this?

**Comment by corinagum** [comment]
Unfortunately it's not scoped out right now. I hope to get to this soon.

**Comment by cwhitten** [comment]
@murcinar thanks for the issue. We take accessibility seriously and are coordinating on a change in the `CardAction` schema definition to include an alt text property for these scenarios. Thanks for your patience as this propagates through the appropriate Bot Framework teams.

**Comment by carlosscastro** [comment]
@corinagum any updates on this one?

**Comment by murcinar** [comment]
Thank you Chris, I really appreciate your help!

**Comment by corinagum** [comment]
Right now we are discussing internally whether or not adding an alt-text to the activity is a viable resolution. The alt-text would be used to add to the image `alt`, so therefore developers can differentiate between what is read by the screenreader if the img conveys information that is not in the button text. 

@murcinar 's use case is a good example, where there is no button text, but the upvote button is commonly used to indicate user response.

I will get an update on the status of this discussion and report back.

**Comment by chrimc62** [comment]
Any comments @cwhitten ?

**Comment by cwhitten** [comment]
Hi @murcinar, we have agreement that the CardAction schema will change to allow alt-text to be provided from the bot. We're still aligning on a date for this to be available. I'll continue to update this issue with new information.

**Comment by murcinar** [comment]
Thank you Chris! @cwhitten

**Comment by cleemullins** [comment]
@murcinar We have (as @cwhitten has already said) the work agreed upon - we just haven't had a chance to book it out yet. I'll update this issue once we assign this to a release.

**Comment by hcyang** [comment]
At the moment, we are still looking to prioritize this issue. This will be at least R10 or after, not for the impending release.

**Comment by corinagum** [comment]
Blocked by SDK changes.

**Comment by compulim** [comment]
https://github.com/microsoft/botframework-sdk/issues/5874 is resolved. We are unblocked now.

**Comment by compulim** [comment]
I think we are not punting this one. As Chris didn't tag it as R11 during our last run.
